x/axelarnet/message_handler.go

```diff
@@ -0,0 +1,282 @@
+
+	coin, err := keeper.NewCoin(ctx, ibcK, n, token)
+	if err != nil {
+		return err
+	}
+
+	if err = coin.Lock(b, types.MessageSender); err != nil {
+		return err
+	}
+
+	transferID, err := n.EnqueueTransfer(ctx, sourceAddress.Chain, crossChainAddr, token)
+	if err != nil {
+		return err
+	}
+
+	ibcK.Logger(ctx).Debug(fmt.Sprintf("enqueued transfer for Message from chain %s", sourceAddress.Chain.Name),
+		"chain", destChain.Name,
+		"transferID", transferID.String(),
+	)
+
+	events.Emit(ctx, &types.TokenSent{
+		TransferID:         transferID,
+		SourceChain:        sourceAddress.Chain.Name,
+		Sender:             sourceAddress.Address,
+		DestinationChain:   nexus.ChainName(msg.DestinationChain),
+		DestinationAddress: msg.DestinationAddress,
+		Asset:              token,
+	})
+
+	return nil
+
+}
+
+// toICS20Packet unmarchal packet to ICS20 token packet
+func toICS20Packet(packet ibcexported.PacketI) (ibctransfertypes.FungibleTokenPacketData, error) {
+	var data ibctransfertypes.FungibleTokenPacketData
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return data, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+	}
+
+	if err := data.ValidateBasic(); err != nil {
+		return data, err
+	}
+
+	return data, nil
+}
+
+// extractTokenFromPacketData get token from ICS20 packet
+// panic if unable to unmarshal packet data
+func extractTokenFromPacketData(packet ibcexported.PacketI) sdk.Coin {
+	var data ibctransfertypes.FungibleTokenPacketData
+	funcs.MustNoErr(types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data))
+
+	// parse the transfer amount
+	amount := funcs.MustOk(sdk.NewIntFromString(data.Amount))
+
+	var denom string
+	if ibctransfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), data.Denom) {
+		// sender chain is not the source, un-escrow token
+
+		// remove prefix added by sender chain
+		voucherPrefix := ibctransfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+		unprefixedDenom := data.Denom[len(voucherPrefix):]
+
+		// coin denomination used in sending from the escrow address
+		denom = unprefixedDenom
+
+		// the denomination used to send the coin is either
+		// -the native denom
+		// -the hash of the path if the denom is not native.
+		denomTrace := ibctransfertypes.ParseDenomTrace(unprefixedDenom)
+		if denomTrace.Path != "" {
+			denom = denomTrace.IBCDenom()
+		}
+	} else {
+		// sender chain is the source
+
+		// since SendPacket did not prefix the denomination, we must prefix denomination here
+		sourcePrefix := ibctransfertypes.GetDenomPrefix(packet.GetDestPort(), packet.GetDestChannel())
+		// NOTE: sourcePrefix contains the trailing "/"
+		prefixedDenom := sourcePrefix + data.Denom
+
+		// construct the denomination trace from the full raw denomination
+		denomTrace := ibctransfertypes.ParseDenomTrace(prefixedDenom)
+
+		denom = denomTrace.IBCDenom()
```

Comment:
we register the base denom in nexus. So the is asset registered check will fail with this value for `uatom` as an e.g. Maybe convert it to the internal coin type first?

Reply:
converted in validate basic

Reply:
i prefer that we return the internal coin version from this method. This is because we want to attach the asset using the nexus naming to the message. We're still attaching the IBC denom version atm.

Reply:
the coin type expect ibc hash to determine if it's a ics20 coin, and we use normalized name when set to general message.